packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts

```diff
@@ -172,47 +172,69 @@ function getEventMap(elm: EventTarget): ListenerMap {
     return listenerInfo;
 }
 
-const shadowRootEventListenerMap: WeakMap<EventListener, WrappedListener> = new WeakMap();
+// Using a WeakMap instead of a WeakSet because this one works in IE11 :(
+const eventsComingFromShadowRoot: WeakMap<Event, 1> = new WeakMap();
+
+function isEventComingFromShadowRoot(event: Event): boolean {
+    return eventsComingFromShadowRoot.has(event);
+}
+
+export function setEventFromShadowRoot(event: Event) {
+    eventsComingFromShadowRoot.set(event, 1);
+}
+
+const shadowRootEventListenerMap: WeakMap<
+    EventListenerOrEventListenerObject,
+    WrappedListener
+> = new WeakMap();
 
 function getWrappedShadowRootListener(
     sr: SyntheticShadowRootInterface,
-    listener: EventListener
+    listener: EventListenerOrEventListenerObject
 ): WrappedListener {
-    if (!isFunction(listener)) {
-        throw new TypeError(); // avoiding problems with non-valid listeners
-    }
-    let shadowRootWrappedListener = shadowRootEventListenerMap.get(listener);
-    if (isUndefined(shadowRootWrappedListener)) {
-        shadowRootWrappedListener = function (event: Event) {
-            // * if the event is dispatched directly on the host, it is not observable from root
-            // * if the event is dispatched in an element that does not belongs to the shadow and it is not composed,
-            //   it is not observable from the root
-            const { composed } = event;
+    let wrappedListener = shadowRootEventListenerMap.get(listener);
+    if (isUndefined(wrappedListener)) {
+        wrappedListener = function (event: Event) {
             const target = eventTargetGetter.call(event);
             const currentTarget = eventCurrentTargetGetter.call(event);
-            if (target !== currentTarget) {
-                const rootNode = getRootNodeHost(
-                    target as Node /* because wrapping on shadowRoot */,
-                    {
-                        composed,
-                    }
-                );
-
-                if (
-                    isChildNode(rootNode as HTMLElement, currentTarget as Node) ||
-                    (composed === false && rootNode === currentTarget)
-                ) {
+            const { composed } = event;
+
+            let invokeListener = false;
+
+            // Confusing! Remember that in synthetic-shadow, roots do not have a physical presence
+            // in the DOM and dispatching an event on a root actually results in the event getting
+            // dispatched on its corresponding host.
+            if (target === currentTarget && !isEventComingFromShadowRoot(event)) {
+                // Do not invoke if the event was dispatched directly on the current root's
+                // corresponding host.
+                invokeListener = false;
```

Comment:
Something looks wrong here, the `invokeListener` variable is already set to false line 202. Why resetting it?

Reply:
I think it's easier to read if we're explicit about what the value, even if we're not actually changing it...especially since it's confusing.

Reply:
What about something like? Let me know if you think it's more readable. Otherwise, let's roll with your approach.

```js
const isDispatchedOnHost = target === currentTarget && !isEventComingFromShadowRoot(event);
if (!isDispatchedOnHost) {
  if (isComposed) {
    // Some code...
  } else {
    // More code...
  }
}
```